src/time/Waveform.hpp

```diff
@@ -0,0 +1,56 @@
+#pragma once
+
+#include <Eigen/Core>
+#include "logging/Logger.hpp"
+
+namespace precice {
+namespace time {
+
+class Waveform {
+public:
+  /**
+   * @brief Waveform object which stores data of current and past time windows for performing extrapolation.
+   * @param numberOfData defines how many pieces of data one sample in time consists of
+   * @param extrapolatioOrder defines the maximum extrapolation order supported by this Waveform and reserves storage correspondingly
+   */
+  Waveform(int numberOfData,
+           int extrapolationOrder);
+
+  /**
+   * @brief Updates entry in _timeWindows corresponding to this window with given data
+   * @param data new sample for this time window
+   */
+  void updateThisWindow(Eigen::VectorXd data);
+
+  /**
+   * @brief Called, when moving to the next time window. All entries in _timeWindows are shifted. The new entry is initialized as the value from the last window (= constant extrapolation)
+   * @param timeWindows number of samples that are valid and may be used for extrapolation. Usually number of past time windows.
+   */
+  void moveToNextWindow(int timeWindows, int order = 0);
```

Comment:
```suggestion
  void moveToNextWindow();
```
Removing the two arguments would simplify the interface easier to use, but the class would also become more complex:

* (extrapolation) `order` is defined when the object is created
* Keep track of `timeWindows` inside object (to some degree a duplication of what the coupling scheme also does.)

@uekerman Any opinion? We could also shift this discussion to the next PR, where interpolation will be introduced.

Reply:
This complexity will anyway be moved to the new package. I don't see a problem doing this now.